### PR TITLE
Integrate physutils - Physio object usage

### PR DESCRIPTION
**Pull request by maestroque**
## Proposed Changes

  - Integrate physutils (currently only testing locally as the package is pending publishing)
  - Integrate the use of the `Physio` object for the metric functions
  - Retain the ability for the functions to be used without the need for the Physio object

## Change Type

- [ ] `bugfix` (+0.0.1)
- [ ] `minor` (+0.1.0)
- [X] `major`  (+1.0.0)
- [ ] `refactoring` (no version update)
- [ ] `test` (no version update)
- [ ] `infrastructure` (no version update)
- [ ] `documentation` (no version update)
- [ ] `other`

## Checklist before review

- [ ] I added everything I wanted to add to this PR.
- [ ] \[Code or tests only\] I wrote/updated the necessary docstrings.
- [ ] \[Code or tests only\] I ran and passed tests locally.
- [ ] \[Documentation only\] I built the docs locally.
- [X] My contribution is harmonious with the rest of the code: I'm not introducing repetitions.
- [X] My code respects the adopted style, especially linting conventions.
- [x] The title of this PR is explanatory on its own, enough to be understood as part of a changelog.
- [ ] I added or indicated the right labels.

- [ ] I added information regarding the timeline of completion for this PR.

- [X] Please, comment on my PR while it's a draft and give me feedback on the development!

### CODE REVIEW

**Comment by maestroque** [comment]
Currently tests are failing, because the `physutils` module cannot be installed

File: phys2denoise/metrics/multimodal.py
Comment (m-miedema) [inline phys2denoise/metrics/multimodal.py:27]: I would suggest including a bit more information here, especially that the object should also contain associated meta-data.

**Review by m-miedema — CHANGES_REQUESTED** [review]
The main thing that I think we need to discuss is checking for peaks/troughs where appropriate in the Physio object. I'm not sure the best way to do this (and we've talked about maintaining functionality for non-Physio object time-series, so we'll need to think about that in this context too). Many parameters e.g. TR could equally be associated with a Physio object, so we need a clear plan that checks for parameters not supplied explicitly but allows flexibility for a non-Physio implementation.

File: phys2denoise/metrics/multimodal.py
Comment (m-miedema) [reply phys2denoise/metrics/multimodal.py:27]: (I think it's worth a discussion where we talk about advantages of forcing some parameters to be in the Physio object vs. explicitly specified before we go too far down this road!)

**Comment by maestroque** [comment]
I believe the support for Physio and non-Physio inputs has been completed. @m-miedema @me-pic let me know your thoughts

The unit tests now need updating.

**Comment by maestroque** [comment]
In CRF, iCRF and RRF there is this line:
```python
time_stamps = np.arange(0, time_length, 1 / samplerate)
```

Since samplerate is in seconds, shouldn't it be
```python
time_stamps = np.arange(0, time_length, samplerate)
```

Do you prefer changing the doc to Hz or is it more convinient in seconds?
@m-miedema @me-pic

**Comment by maestroque** [comment]
Also in general, shouldn't we opt in a single way to define sample rate? Is there any reason to not have everything in Hz?

File: phys2denoise/tests/test_metrics_chest_belt.py
Comment (maestroque) [inline phys2denoise/tests/test_metrics_chest_belt.py:27]: Not sure why this was marked as xfail, but now it's fixed and it passes.

**Comment by maestroque** [comment]
@smoia I think `style_check` in circleci fails because the style_check image python version is 3.7 which does not support matplotlib >=3.9, is that correct?

Should I change the version [here](https://github.com/physiopy/phys2denoise/blob/master/.circleci/config.yml#L88)?

**Comment by maestroque** [comment]
Ok, now I understand, so we want to be compatible still with python 3.7 right? if that's the case, we cannot have matplotlib >=3.9.
`style_check` now passes, `test37` fails due to the aforementioned dep conflict, and `test310` fails only due to the numpy version change, which makes certain unit tests fail locally as well.

I guess we should decide if we want to stay compatible with python 3.7 prior to merging this PR. The `numpy` deprecation is being handled separately correct?
@smoia @m-miedema @me-pic

**Comment by smoia** [comment]
Hey @maestroque ! Would you be able to update our CircleCI tests to support 3.8 and 3.12 rather tan 3.7 and 3.10 respectively?

**Comment by smoia** [comment]
> Should I change the version [here](https://github.com/physiopy/phys2denoise/blob/master/.circleci/config.yml#L88)?

And yes, correct! That's one of the places to change versions - any other place will look like that as well. Don't forget to increase the python required by the `setup.cfg` as well, and before we forget, I'll slam a breaking label on this release.

**Comment by maestroque** [comment]
Perfect, that sounds good!

**Comment by maestroque** [comment]
@smoia actually I see that python 3.8 caps matplotlib at 3.7.5. Shall I change to python 3.9?

**Comment by maestroque** [comment]
Ok, now with 3.9 and 3.12:
- The 3.9 fails are the ones stemming from numpy
- 3.12 fails because versioneer has had issues with python 3.12 [[note]](https://github.com/pydata/pandas-datareader/issues/969)
From versioneer README
> Experimental support for Python 3.12.

@smoia

**Comment by maestroque** [comment]
Ok fails are now the expected ones, for both 3.9 and 3.11 tests

**Comment by maestroque** [comment]
I added an extra feature in this PR, that will hopefully be useful in the workflow.

I opted to store the computed metrics inside the Physio object, in order to make exporting them much easier in the workflow. E.g. having a task to compute/visualize and a task to export

Wdyt? @m-miedema @me-pic @smoia

**Comment by m-miedema** [comment]
> I added an extra feature in this PR, that will hopefully be useful in the workflow.
> 
> I opted to store the computed metrics inside the Physio object, in order to make exporting them much easier in the workflow. E.g. having a task to compute/visualize and a task to export
> 
> Wdyt? @m-miedema @me-pic @smoia

Good idea! Do you have plans to follow through with this in the history field?

**Comment by maestroque** [comment]
@m-miedema metrics calculations are already integrated with the `history`. This new attribute is to store all the metric values apart from only the history. This attribute can be also used in `physioqc` for example

**Comment by maestroque** [comment]
Currently blocked by https://github.com/physiopy/physutils/pull/4, due to important physio object fix (release needed): [0c324c5](https://github.com/physiopy/physutils/pull/4/commits/0c324c5c04f4142160bc9a0705110cafe67d9473)

**Comment by maestroque** [comment]
@smoia @me-pic @m-miedema I added a wrapper to sort out the returns (only metric or physio with metric), feel free to review.
The failing tests currently are due to:
- physutils package not being up to date with https://github.com/physiopy/physutils/pull/4, where I pushed some bug fixes
- Existing fail in master due to a failed error assertion. The error is indeed raised but the assertion does not catch it.

Also, the following changes were made to deps:
- numpy: >=1.9.3, <2
- matplotlib: non-constrained

File: phys2denoise/tests/test_metrics_cardiac.py
Comment (maestroque) [inline phys2denoise/tests/test_metrics_cardiac.py:62]: Maybe it would be beneficial to add a `Metric` class/struct?

**Comment by maestroque** [comment]
I believe the Metric class will make it much easier for the pydra tasks to handle computing and exporting metrics

File: .all-contributorsrc
Comment (smoia) [inline .all-contributorsrc:121]:
```suggestion
        "ideas",
        "infra",
```

File: setup.cfg
Comment (smoia) [inline setup.cfg:24]: Can you open an issue to state we need to refactor code to handle numpy >=2.0?

File: setup.cfg
Comment (smoia) [inline setup.cfg:53]: With the physutils integration, do we still need peakdet?

Comment (maestroque) [reply setup.cfg:53]: yes, there is a unit test with peak finding

File: phys2denoise/metrics/utils.py
Comment (smoia) [inline phys2denoise/metrics/utils.py:356]: We did not implemented it yet, but technically all metrics could have lags - and all metrics could be convolved. Is this a way to check the type of output you get?

File: phys2denoise/metrics/chest_belt.py
Comment (smoia) [inline phys2denoise/metrics/chest_belt.py:197]: Now that you can either input a physio object or a numpy array, do we still need to duplicate code for this application?

File: phys2denoise/metrics/chest_belt.py
Comment (smoia) [reply phys2denoise/metrics/chest_belt.py:19]: It would be good to have peaks and troughs first when possible (and, in general, avoid changing the order of inputs as much as possible), but if it makes things too complicated with the wrapper, then no problem.

File: phys2denoise/metrics/utils.py
Comment (smoia) [reply phys2denoise/metrics/utils.py:356]: Could you instead make a check on the provided arguments to the function, to check whether convolution and/or lag happened?

File: phys2denoise/metrics/chest_belt.py
Comment (smoia) [inline phys2denoise/metrics/chest_belt.py:28]: Could we change this to `physutils.Physio, np.ndarray, or array-like object` instead?
That's also possible now, right?

File: phys2denoise/metrics/chest_belt.py
Comment (smoia) [inline phys2denoise/metrics/chest_belt.py:50]: I was that you repeat this few following lines often in different metrics - if that is the case, I would suggest to make it a function on its own or even better to add it to the decorator you created. If you make it a function, the return can be done within the metric function itself (without the wrapper), if you move this to the decorator (better), you can revert to the original metric functions, and deal with the input and returns of the metrics within the wrapper!

**Review by smoia — COMMENTED** [review]
Couple of notes for the moment!

File: setup.cfg
Comment (maestroque) [reply setup.cfg:24]: Opened here: #62

File: phys2denoise/metrics/utils.py
Comment (maestroque) [reply phys2denoise/metrics/utils.py:356]: I'll change to detect lagged metrics, since it's dependent on the argument. However convolution is now only done in some metrics without being specified in the arguments and can't be detected easily if you don't read the code. Also you can't deduce that the metric is convolved from the dimensions of the metric output.

I can open an issue about it if you want

File: phys2denoise/metrics/chest_belt.py
Comment (maestroque) [reply phys2denoise/metrics/chest_belt.py]: yeap, done

File: phys2denoise/metrics/chest_belt.py
Comment (maestroque) [reply phys2denoise/metrics/chest_belt.py:197]: That's true the relevant line in `env()` could be changed as such:
```python
env_arr = (
        pd.Series(data.data)
        .rolling(window=window, center=True)
        .apply(respiratory_pattern_variability, args=(window,), kwargs=dict(return_physio=False))
    )
```
However we should also include support for `pandas.core.series.Series` types in the `load_physio` function for that to work, which also impacts the other module. Still, I think it'd be better to open an issue about it, because it complicates the PRs
Wdyty @smoia

File: phys2denoise/metrics/chest_belt.py
Comment (maestroque) [reply phys2denoise/metrics/chest_belt.py:19]: done

**Comment by maestroque** [comment]
@smoia please recheck when you find some time

File: phys2denoise/tests/test_metrics_utils.py
Comment (m-miedema) [inline phys2denoise/tests/test_metrics_utils.py:23]: I get the expected logger warning here, so the test fails. I'm guessing this one is what you were referring to as the failed catch [here](https://github.com/physiopy/phys2denoise/pull/54#issuecomment-2282224388)? If so, I'm wondering what your suggestion is for handling it to catch the warning instead.

**Review by m-miedema — APPROVED** [review]
With the updated physutils this looks good to me! I don't see a reason not to fix the one test I mentioned on the way out but (feedback from @smoia pending) I would be happy to see this move on and into the workflow.

File: phys2denoise/tests/test_metrics_utils.py
Comment (maestroque) [reply phys2denoise/tests/test_metrics_utils.py:23]: Yes that's the one. It's not detected because it's being handled. I'll just add a check that the subsequent warning is raised

Comment (maestroque) [reply phys2denoise/tests/test_metrics_utils.py:23]: It's a bit convoluted though and it's pending a full integration of loguru in this package as well. I'd say we merge and document in the issue/PR #53

Comment (maestroque) [reply phys2denoise/tests/test_metrics_utils.py]: I did a quick fix and it detects the expected warning text (as in the peakdet warning assertion tests I did), however the full integration of loguru needs to be adressed: [Relevant note](https://github.com/physiopy/phys2denoise/issues/48#issuecomment-2308829082)

@m-miedema

**Comment by maestroque** [comment]
Now, CI passes online as well, because of the `physutils` release. I'm merging @m-miedema @smoia

File: phys2denoise/tests/test_metrics_utils.py
Comment (m-miedema) [reply phys2denoise/tests/test_metrics_utils.py]: Thanks @maestroque !

**Pull request merged**
